// @opengovsg/design-system-react/build/main/Sidebar/SidebarContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSidebarContext = exports.SidebarProvider = exports.useSidebarNestContext = exports.SidebarNestProvider = exports.useSidebarStyles = exports.SidebarStylesProvider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const react_2 = require("@chakra-ui/react");
const [SidebarStylesProvider, useSidebarStyles] = (0, react_2.createStylesContext)('Sidebar');
exports.SidebarStylesProvider = SidebarStylesProvider;
exports.useSidebarStyles = useSidebarStyles;
const SidebarNestContext = (0, react_1.createContext)(undefined);
const SidebarNestProvider = ({ children, nested }) => {
    return ((0, jsx_runtime_1.jsx)(SidebarNestContext.Provider, { value: { nested }, children: children }));
};
exports.SidebarNestProvider = SidebarNestProvider;
const useSidebarNestContext = () => {
    const context = (0, react_1.useContext)(SidebarNestContext);
    if (!context) {
        return {
            nested: false,
        };
    }
    return context;
};
exports.useSidebarNestContext = useSidebarNestContext;
const SidebarContext = (0, react_1.createContext)(undefined);
const SidebarProvider = ({ children, reduceMotion = false, }) => {
    return ((0, jsx_runtime_1.jsx)(SidebarContext.Provider, { value: { reduceMotion }, children: children }));
};
exports.SidebarProvider = SidebarProvider;
const useSidebarContext = () => {
    const context = (0, react_1.useContext)(SidebarContext);
    if (!context) {
        return {
            reduceMotion: false,
        };
    }
    return context;
};
exports.useSidebarContext = useSidebarContext;
