// @opengovsg/design-system-react/build/main/theme/components/CloseButton.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloseButton = void 0;
const react_1 = require("@chakra-ui/react");
const theme_tools_1 = require("@chakra-ui/theme-tools");
const layerStyles_1 = require("../layerStyles");
const Button_1 = require("./Button");
const $size = (0, theme_tools_1.cssVar)('close-button-size');
const sizes = {
    lg: (0, react_1.defineStyle)({
        [$size.variable]: (_a = Button_1.Button.sizes) === null || _a === void 0 ? void 0 : _a.lg.minH,
    }),
    md: (0, react_1.defineStyle)({
        [$size.variable]: (_b = Button_1.Button.sizes) === null || _b === void 0 ? void 0 : _b.md.minH,
    }),
    sm: (0, react_1.defineStyle)({
        [$size.variable]: (_c = Button_1.Button.sizes) === null || _c === void 0 ? void 0 : _c.sm.minH,
    }),
    xs: (0, react_1.defineStyle)({
        [$size.variable]: (_d = Button_1.Button.sizes) === null || _d === void 0 ? void 0 : _d.xs.minH,
    }),
};
const baseStyle = (0, react_1.defineStyle)({
    p: 0,
    _focusVisible: layerStyles_1.layerStyles.focusRing.default._focusVisible,
    w: [$size.reference],
    h: [$size.reference],
});
const variantClear = (0, react_1.defineStyle)((props) => {
    var _a;
    return Object.assign(Object.assign({}, (_a = Button_1.Button.variants) === null || _a === void 0 ? void 0 : _a.clear(props)), { px: 'initial' });
});
const variants = {
    clear: variantClear,
};
exports.CloseButton = (0, react_1.defineStyleConfig)({
    baseStyle,
    sizes,
    variants,
    defaultProps: {
        colorScheme: 'neutral',
        size: 'md',
    },
});
