// @opengovsg/design-system-react/build/main/theme/components/DatePicker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatePicker = exports.datepickerAnatomy = void 0;
const react_1 = require("@chakra-ui/react");
const theme_tools_1 = require("@chakra-ui/theme-tools");
const textStyles_1 = require("../textStyles");
exports.datepickerAnatomy = (0, theme_tools_1.anatomy)('datepicker').parts('header', 'inputButton');
const { defineMultiStyleConfig, definePartsStyle } = (0, react_1.createMultiStyleConfigHelpers)(exports.datepickerAnatomy.keys);
const sizes = {
    xs: definePartsStyle({
        inputButton: {
            fontSize: '1rem',
        },
    }),
    sm: definePartsStyle({
        inputButton: {
            fontSize: '1.25rem',
        },
    }),
    md: definePartsStyle({
        inputButton: {
            fontSize: '1.25rem',
        },
    }),
};
exports.DatePicker = defineMultiStyleConfig({
    baseStyle: {
        header: Object.assign({ h: '3.5rem', display: 'flex', justifyContent: 'space-between', alignItems: 'center', color: 'base.content.strong', px: '1.5rem' }, textStyles_1.textStyles['subhead-2']),
        inputButton: {
            color: 'base.content.strong',
            _groupDisabled: {
                color: 'interaction.support.disabled-content',
            },
        },
    },
    sizes,
});
