// @opengovsg/design-system-react/build/main/theme/components/Divider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Divider = void 0;
const react_1 = require("@chakra-ui/react");
const baseStyle = (0, react_1.defineStyle)({
    opacity: 1,
    borderStyle: 'solid',
    _dark: {
        borderColor: 'base.divider.inverse',
    },
});
const variantLight = (0, react_1.defineStyle)({
    borderColor: 'base.divider.subtle',
    _dark: {
        borderColor: 'base.divider.inverse',
    },
    borderWidth: '1px',
});
const variantMedium = (0, react_1.defineStyle)({
    borderColor: 'base.divider.medium',
    _dark: {
        borderColor: 'base.divider.inverse',
    },
    borderWidth: '1px',
});
const variantThick = (0, react_1.defineStyle)({
    borderColor: 'base.divider.strong',
    _dark: {
        borderColor: 'base.divider.inverse',
    },
    borderWidth: '2px',
});
const variantBrand = (0, react_1.defineStyle)({
    borderColor: 'base.divider.brand',
    _dark: {
        borderColor: 'base.divider.inverse',
    },
    borderWidth: '2px',
});
const variantInverse = (0, react_1.defineStyle)({
    borderColor: 'base.divider.inverse',
    borderWidth: '2px',
});
const variants = {
    light: variantLight,
    medium: variantMedium,
    thick: variantThick,
    brand: variantBrand,
    inverse: variantInverse,
};
exports.Divider = (0, react_1.defineStyleConfig)({
    baseStyle,
    variants,
    defaultProps: {
        variant: 'medium',
    },
});
