// @opengovsg/design-system-react/build/main/theme/components/FormControl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormControl = void 0;
const anatomy_1 = require("@chakra-ui/anatomy");
const styled_system_1 = require("@chakra-ui/styled-system");
const textStyles_1 = require("../textStyles");
const parts = anatomy_1.formAnatomy.extend('icon', 'optionalIndicator');
const { definePartsStyle, defineMultiStyleConfig } = (0, styled_system_1.createMultiStyleConfigHelpers)(parts.keys);
const $fg = (0, styled_system_1.cssVar)('form-control-color');
const baseStyleRequiredIndicator = (0, styled_system_1.defineStyle)({
    marginStart: '1',
    [$fg.variable]: 'colors.utility.feedback.critical',
    color: $fg.reference,
});
const baseStyleHelperText = (0, styled_system_1.defineStyle)(Object.assign(Object.assign({ [$fg.variable]: 'colors.base.content.medium', color: $fg.reference }, textStyles_1.textStyles['body-2']), { mt: 0 }));
const baseStyleOptionalIndicator = (0, styled_system_1.defineStyle)(Object.assign(Object.assign({ color: 'base.content.medium' }, textStyles_1.textStyles['body-2']), { ml: '0.5rem', lineHeight: 0 }));
const baseStyle = definePartsStyle({
    requiredIndicator: baseStyleRequiredIndicator,
    helperText: baseStyleHelperText,
    optionalIndicator: baseStyleOptionalIndicator,
    icon: {
        marginEnd: '0.5em',
        fontSize: '1rem',
        height: '1.25em',
        verticalAlign: 'bottom',
    },
});
const variantSuccess = definePartsStyle({
    helperText: {
        [$fg.variable]: 'colors.utility.feedback.success',
        color: $fg.reference,
    },
});
const variants = {
    success: variantSuccess,
};
exports.FormControl = defineMultiStyleConfig({
    baseStyle,
    variants,
    defaultProps: {
        size: 'md',
    },
});
