// @opengovsg/design-system-react/build/main/theme/components/Infobox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Infobox = void 0;
const react_1 = require("@chakra-ui/react");
const theme_tools_1 = require("@chakra-ui/theme-tools");
const parts = (0, theme_tools_1.anatomy)('infobox').parts('messagebox', 'icon');
const { definePartsStyle, defineMultiStyleConfig } = (0, react_1.createMultiStyleConfigHelpers)(parts.keys);
const baseStyle = definePartsStyle({
    messagebox: {
        display: 'flex',
        justifyContent: 'start',
        color: 'base.content.strong',
    },
});
const variantInfo = definePartsStyle({
    messagebox: {
        bg: 'utility.feedback.info-subtle',
    },
    icon: {
        color: 'utility.feedback.info',
    },
});
const variantWarning = definePartsStyle({
    messagebox: {
        bg: 'utility.feedback.warning-subtle',
    },
    icon: {
        color: 'utility.feedback.warning',
    },
});
const variantError = definePartsStyle({
    messagebox: {
        bg: 'utility.feedback.critical-subtle',
    },
    icon: {
        color: 'utility.feedback.critical',
    },
});
const sizes = {
    sm: definePartsStyle({
        messagebox: {
            p: '0.625rem',
            textStyle: 'body-2',
        },
        icon: {
            my: '0.125rem',
            fontSize: '1rem',
            mr: '0.5rem',
        },
    }),
    md: definePartsStyle({
        messagebox: {
            padding: '1rem',
            textStyle: 'body-1',
        },
        icon: {
            fontSize: '1.5rem',
            mr: '0.5rem',
        },
    }),
};
const variants = {
    info: variantInfo,
    warning: variantWarning,
    error: variantError,
};
exports.Infobox = defineMultiStyleConfig({
    baseStyle,
    variants,
    sizes,
    defaultProps: {
        variant: 'info',
        size: 'md',
    },
});
