// @opengovsg/design-system-react/build/main/theme/components/Progress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Progress = void 0;
const anatomy_1 = require("@chakra-ui/anatomy");
const theme_tools_1 = require("@chakra-ui/theme-tools");
const baseStyleTrack = ({ colorScheme: c }) => {
    return {
        bg: `${c}.200`,
        borderRadius: '1.25rem',
    };
};
const filledStyle = ({ colorScheme: c, theme, isIndeterminate, }) => {
    const bgColor = `${c}.500`;
    const gradient = `linear-gradient(
    to right,
    transparent 0%,
    ${(0, theme_tools_1.getColor)(theme, bgColor)} 50%,
    transparent 100%
  )`;
    return Object.assign({}, (isIndeterminate ? { bgImage: gradient } : { bgColor }));
};
const baseStyleFilledTrack = (props) => {
    return Object.assign({ transitionProperty: 'all', transitionDuration: 'slow' }, filledStyle(props));
};
const baseStyle = (props) => ({
    track: baseStyleTrack(props),
    filledTrack: baseStyleFilledTrack(props),
});
const sizes = {
    xl: {
        track: { h: '1.5rem' },
    },
};
exports.Progress = {
    baseStyle,
    parts: anatomy_1.progressAnatomy.keys,
    sizes,
    defaultProps: {
        size: 'md',
        colorScheme: 'brand.primary',
    },
};
