// @opengovsg/design-system-react/build/main/theme/components/Searchbar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Searchbar = void 0;
const react_1 = require("@chakra-ui/react");
const theme_tools_1 = require("@chakra-ui/theme-tools");
const parts = (0, theme_tools_1.anatomy)('searchbar').parts('icon');
const { definePartsStyle, defineMultiStyleConfig } = (0, react_1.createMultiStyleConfigHelpers)(parts.keys);
const variantOutline = definePartsStyle({
    icon: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        color: 'base.content.strong',
    },
});
const sizes = {
    xs: definePartsStyle({
        icon: {
            fontSize: '0.875rem',
        },
    }),
    sm: definePartsStyle({
        icon: {
            fontSize: '1rem',
        },
    }),
    md: definePartsStyle({
        icon: {
            fontSize: '1.25rem',
        },
    }),
};
const variants = {
    outline: variantOutline,
};
exports.Searchbar = defineMultiStyleConfig({
    variants,
    sizes,
    defaultProps: {
        variant: 'outline',
        size: 'md',
    },
});
