// @opengovsg/design-system-react/build/main/theme/components/Textarea.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Textarea = void 0;
const react_1 = require("@chakra-ui/react");
const Input_1 = require("./Input");
const sizes = {
    xs: (0, react_1.defineStyle)(Object.assign(Object.assign({}, (_a = Input_1.Input.sizes) === null || _a === void 0 ? void 0 : _a.xs.field), { px: '0.75rem', py: '0.5rem' })),
    sm: (0, react_1.defineStyle)(Object.assign(Object.assign({}, (_b = Input_1.Input.sizes) === null || _b === void 0 ? void 0 : _b.sm.field), { px: '0.75rem', py: '0.625rem' })),
    md: (0, react_1.defineStyle)(Object.assign(Object.assign({}, (_c = Input_1.Input.sizes) === null || _c === void 0 ? void 0 : _c.md.field), { px: '1rem', py: '0.625rem' })),
};
const baseStyle = (0, react_1.defineStyle)({
    transitionProperty: 'common',
    transitionDuration: 'normal',
});
const variantOutline = (0, react_1.defineStyle)((props) => {
    var _a;
    return (Object.assign({}, (_a = Input_1.Input.variants) === null || _a === void 0 ? void 0 : _a.outline(props).field));
});
const variants = {
    outline: variantOutline,
};
exports.Textarea = (0, react_1.defineStyleConfig)({
    baseStyle,
    sizes,
    variants,
    defaultProps: Input_1.Input.defaultProps,
});
