// @opengovsg/design-system-react/build/main/SingleSelect/utils/itemUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.itemToDescriptionString = exports.isItemDisabled = exports.itemToIcon = exports.itemToLabelString = exports.itemToValue = exports.itemIsObject = void 0;
const itemIsObject = (item) => {
    return !!item && typeof item !== 'string';
};
exports.itemIsObject = itemIsObject;
const itemToValue = (item) => {
    if (!item)
        return '';
    if (!(0, exports.itemIsObject)(item)) {
        return item;
    }
    return item.value;
};
exports.itemToValue = itemToValue;
const itemToLabelString = (item) => {
    var _a;
    if (!item)
        return '';
    if (!(0, exports.itemIsObject)(item)) {
        return item;
    }
    return (_a = item.label) !== null && _a !== void 0 ? _a : item.value;
};
exports.itemToLabelString = itemToLabelString;
const itemToIcon = (item) => {
    if (!item || !(0, exports.itemIsObject)(item)) {
        return undefined;
    }
    return item.icon;
};
exports.itemToIcon = itemToIcon;
const isItemDisabled = (item) => {
    return (0, exports.itemIsObject)(item) && !!item.disabled;
};
exports.isItemDisabled = isItemDisabled;
const itemToDescriptionString = (item) => {
    return (0, exports.itemIsObject)(item) ? item.description : undefined;
};
exports.itemToDescriptionString = itemToDescriptionString;
