// @opengovsg/design-system-react/build/main/theme/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.theme = void 0;
const react_1 = require("@chakra-ui/react");
const breakpoints_1 = require("./foundations/breakpoints");
const colours_1 = require("./foundations/colours");
const semanticTokens_1 = require("./foundations/semanticTokens");
const shadows_1 = require("./foundations/shadows");
const spacing_1 = require("./foundations/spacing");
const transition_1 = require("./foundations/transition");
const typography_1 = require("./foundations/typography");
const components_1 = require("./components");
const layerStyles_1 = require("./layerStyles");
const textStyles_1 = require("./textStyles");
exports.theme = (0, react_1.extendTheme)({
    styles: {
        global: {
            body: {
                fontFeatureSettings: "'tnum' on, 'cv05' on",
            },
        },
    },
    semanticTokens: semanticTokens_1.semanticTokens,
    colors: colours_1.colours,
    breakpoints: breakpoints_1.breakpoints,
    shadows: shadows_1.shadows,
    space: spacing_1.spacing,
    transition: transition_1.transition,
    fonts: {
        heading: 'Inter,Trebuchet MS,-apple-system,Arial,BlinkMacSystemFont,sans-serif',
        body: 'Inter,Trebuchet MS,-apple-system,Arial,BlinkMacSystemFont,sans-serif',
        code: 'IBM Plex Mono,Courier,Monaco,Courier New,monospace',
    },
    fontSizes: typography_1.typography.fontSize,
    fontWeights: typography_1.typography.fontWeights,
    lineHeights: typography_1.typography.lineHeights,
    letterSpacings: typography_1.typography.letterSpacing,
    textStyles: textStyles_1.textStyles,
    layerStyles: layerStyles_1.layerStyles,
    components: components_1.components,
});
